### ISSUE TITLE
Add ARM static binary in the release

### ISSUE THREAD

**Issue by meox** [iovisor/bpftrace#2868]
I'd like to have a way to download a static image or a docker image of bpftrace for ARM64

**Comment by sanmai-NL** [comment]
Yes. A tricky thing is that this copy not match the target `--platform` of a container build if it's `aarch64`:

```Dockerfile
COPY --from=quay.io/iovisor/bpftrace /usr/bin/bpftrace /usr/bin/
```

This product's Nix flake does support aarch64: https://github.com/iovisor/bpftrace/blob/289f0c7e028b2317aeddbb0228d542f6250fe08c/flake.nix#L21

Alpine Linux does have an `aarch64` package: https://pkgs.alpinelinux.org/packages?name=bpftrace&branch=edge&repo=&arch=aarch64&maintainer=.

**Comment by danobi** [comment]
I would say it's easy :p. One of the deps we override triggers a systemd rebuild so it would probably take like 12 hours to do that under emulation

**Comment by meox** [comment]
Thanks for your suggestions:

In a Debian ARM64 I installed asciidoctor and nix,
I also changed the ./scripts/create-assets.sh in order to enable nix-command and flakes
After that I run the script (as root due to some permissions issue) and I got the artifact

![Screenshot 2023-12-05 at 18 00 55](https://github.com/iovisor/bpftrace/assets/852548/21b7aa1b-09c6-4a95-b46a-a4ea93172d29)

**Comment by sanmai-NL** [comment]
> I wouldn't say it's easy :p. One of the deps we override triggers a systemd rebuild so it would probably take like 12 hours to do that under emulation.
> 
> 
> 
> Also that command doesn't seem to work for me:
> 
> ```
> 
> $ docker run --arch=arm64 --rm -it alpine:3.18
> 
> unknown flag: --arch
> 
> See 'docker run --help'.
> 
> ```

Sorry, `podman` provides `--arch`, `docker` only `--platform`.

Yeah, I assumed the Nix build was very efficient but if not it could take a while.

**Comment by sanmai-NL** [comment]
> Thanks for your suggestions:
> 
> 
> 
> In a Debian ARM64 I installed asciidoctor and nix,
> 
> I also changed the ./scripts/create-assets.sh in order to enable nix-command and flakes
> 
> After that I run the script (as root due to some permissions issue) and I got the artifact
> 
> 
> 
> ![Screenshot 2023-12-05 at 18 00 55](https://github.com/iovisor/bpftrace/assets/852548/21b7aa1b-09c6-4a95-b46a-a4ea93172d29)
> 

Any information on elapsed time? Not 12 hours I suppose. But you're running it natively.

**Comment by meox** [comment]
> > Thanks for your suggestions:
> > In a Debian ARM64 I installed asciidoctor and nix,
> > I also changed the ./scripts/create-assets.sh in order to enable nix-command and flakes
> > After that I run the script (as root due to some permissions issue) and I got the artifact
> > ![Screenshot 2023-12-05 at 18 00 55](https://private-user-images.githubusercontent.com/852548/288138490-21b7aa1b-09c6-4a95-b46a-a4ea93172d29.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE3OTg3OTIsIm5iZiI6MTcwMTc5ODQ5MiwicGF0aCI6Ii84NTI1NDgvMjg4MTM4NDkwLTIxYjdhYTFiLTA5YzYtNGE5NS1iNDZhLWE0ZWE5MzE3MmQyOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIwNVQxNzQ4MTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wZGJhNzQ1ODBkYzM2YWYwZjc3NzgwODRkMGQ5MDU3OTI3M2RkZWU4ZjIyN2IzMzNmZDg5ZDVhYTM0MzNjMTNhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.mhDraKMlP2ImvK7PWlG_Tt6T2GThujqaRMlo-jydv6Q)
> 
> Any information on elapsed time? Not 12 hours I suppose. But you're running it natively (Mac M*?)?

more or less 10min

**Comment by meox** [comment]
Sorry yes I have a VM Debian running on qemu on a MAC M1 PRO

**Comment by meox** [comment]
Do you need more test from my side?

**Comment by danobi** [comment]
Hi @meox , sorry I forgot to respond :p. Nope, lgtm. I'll put it on my todo list to upload arm binaries -- it'll take a little tweaking in the GH workflow to fan out to arm/x86 hosts to build the arm binary and then fan back in to upload in one go.

**Comment by jeez** [comment]
+1 to having arm and arm64 static binaries available on each release, please. Just FYI and in case it helps, I've been using https://github.com/facebookexperimental/ExtendedAndroidTools for building bpftrace static binaries for these architectures for a while now. It's maintained by @michalgr and it works great. e.g. it's as simple as:

$ ./scripts/build-docker-image.sh && ./scripts/run-docker-build-env.sh
> make bpftrace STATIC_LINKING=true LLVM_BPF_ONLY=true NDK_ARCH=arm64 BUILD_TYPE=Release

**Comment by danobi** [comment]
I took a closer look and it seems like we cannot get access to an ARM runner without paying one way or another:

* https://github.blog/changelog/2023-10-30-accelerate-your-ci-cd-with-arm-based-hosted-runners-in-github-actions/
* https://github.blog/2023-10-02-introducing-the-new-apple-silicon-powered-m1-macos-larger-runner-for-github-actions/

TBH, if we're going to pay money, I'd prefer the project use something like https://actuated.dev/. We want to run vmtests in CI anyways (run our test suite in a VM so we have a matrix of kernel versions) which requires nested virtualization (which requires $$$ large runners).

The TLDR is that we probably need funding to do this (and other nice things) automatically (see https://opencollective.com/bpftrace). But maybe for now I can go buy a raspberry pi and manually do uploads

**Comment by sanmai-NL** [comment]
If you qualify as OSS you can get a free M1 Runner on GitLab.com. https://docs.gitlab.com/ee/ci/runners/saas/macos_saas_runner.html

**Comment by viktormalik** [comment]
Let me check if I can get hands on an ARM machine for this purpose.

**Comment by meox** [comment]
any update?

**Comment by viktormalik** [comment]
Sorry for the delay. I recently got a Raspberry Pi 4 for experimenting so I'll try to use it for building static ARM binaries.

**Comment by JiaHuann** [comment]
Hi~ Do we have arm64 static package now? :D

**Comment by viktormalik** [comment]
Damn, this slipped of my radar. I'm off the next week but will have a look the week after.

**Comment by caizixian** [comment]
It would indeed be very convenient to have a statically linked binary since the distro package can often be behind, and dynamically links against distro's preferred toolchain.

**Comment by JiaHuann** [comment]
> It would indeed be very convenient to have a statically linked binary since the distro package can often be behind, and dynamically links against distro's preferred toolchain.
> 
> 

BTW，I am curious about how to get a static release for the android or what else release. thanks.

**Comment by Ablu** [comment]
> In theory you should be able to `nix build .#appimage` on aarch64 and get a "static" binary (technically an appimage): https://github.com/iovisor/bpftrace/blob/master/docs/nix.md#build-bpftrace-as-a-statically-linked-binary
> 
> But I don't have an aarch64 machine handy. Mind giving it a try?

I can confirm that this works great!

I uploaded a random build here: https://fileserver.linaro.org/s/oCmoKgbEaLWY5AJ
(no warranties and I am not planning to update it regulary, but it might help people to verify everything works as intended).

**Comment by danobi** [comment]
Sweet. Thanks for the heads up.

I've added some notes about overhauling the release assets: https://github.com/bpftrace/bpftrace/issues/3739.

FWIW this is somewhat low on my priority list, so help would be welcome :)

**Comment by danobi** [comment]
Probably need to refactor `scripts/create-assets.sh` a little bit. Then fan-out/fan-in build steps in: https://github.com/bpftrace/bpftrace/blob/master/.github/workflows/release.yml

Some hints on fan-out/fan-in:
* https://stackoverflow.com/questions/57498605/github-actions-share-workspace-artifacts-between-jobs
* https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/using-jobs-in-a-workflow#defining-prerequisite-jobs

**Comment by danobi** [comment]
Not quite part of the release, but now we'll have aarch64 builds on master: https://github.com/bpftrace/bpftrace/pull/3795. The artifact is uploaded, so it's basically as good as a release artifact if you're not trying to do a headless download or something. I believe you gotta be logged in to download a GHA artifact.

**Comment by hunanchenxingyu** [comment]
> +1 to having arm and arm64 static binaries available on each release, please. Just FYI and in case it helps, I've been using https://github.com/facebookexperimental/ExtendedAndroidTools for building bpftrace static binaries for these architectures for a while now. It's maintained by [@michalgr](https://github.com/michalgr) and it works great. e.g. it's as simple as:
> 
> > $ ./scripts/build-docker-image.sh && ./scripts/run-docker-build-env.sh
> > $ make bpftrace STATIC_LINKING=true LLVM_BPF_ONLY=true NDK_ARCH=arm64 BUILD_TYPE=Release

I compiled bpftrace according to the current method, but it is dynamically linked, as shown below:

$ file out/android/arm64/bin/bpftrace 
out/android/arm64/bin/bpftrace: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /system/bin/linker64, BuildID[sha1]=cce10138c4fd29d85c8a0563f76541ae8977cfb5, stripped

Dynamic section at offset 0x4b25e98 contains 26 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [libdl.so]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so]
 0x0000000000000001 (NEEDED)             Shared library: [libz.so]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so]
 0x000000000000001e (FLAGS)              BIND_NOW
 0x000000006ffffffb (FLAGS_1)            Flags: NOW PIE
 0x0000000000000015 (DEBUG)              0x0
 0x0000000000000007 (RELA)               0x3c48
 0x0000000000000008 (RELASZ)             6283152 (bytes)
 0x0000000000000009 (RELAENT)            24 (bytes)
 0x000000006ffffff9 (RELACOUNT)          261771
 0x0000000000000017 (JMPREL)             0x601bd8
 0x0000000000000002 (PLTRELSZ)           6672 (bytes)
 0x0000000000000003 (PLTGOT)             0x4b35140
 0x0000000000000014 (PLTREL)             RELA
 0x0000000000000006 (SYMTAB)             0x3b8
 0x000000000000000b (SYMENT)             24 (bytes)
 0x0000000000000005 (STRTAB)             0x2d08
 0x000000000000000a (STRSZ)              3898 (bytes)
 0x000000006ffffef5 (GNU_HASH)           0x2ae0
 0x0000000000000019 (INIT_ARRAY)         0x4b2d0e8
 0x000000000000001b (INIT_ARRAYSZ)       3504 (bytes)
 0x000000006ffffff0 (VERSYM)             0x2758
 0x000000006ffffffe (VERNEED)            0x2a50
 0x000000006fffffff (VERNEEDNUM)         3
 0x0000000000000000 (NULL)               0x0
